### DISCUSSION TITLE
How to get params serialized as object, when argument is null?

### DISCUSSION THREAD

**Discussion by prodigy** [microsoft/vs-streamjsonrpc#1029] [Q&A]
I am writing a client for the deribit.com JSON-RPC server. It expects params to always be an object. They do not use positional parameters. So when sending a request for a method that does not expect any parameters, StreamJsonRpc currently serializes as an empty array. But this server requires it to be an empty object. How would I achieve that using StreamJsonRpc?

```
trace Information: 8 : {"id":2,"method":"public/get_time"}
trace Verbose: 8 : Sent: {
  "jsonrpc": "2.0",
  "id": 2,
  "method": "public/get_time",
  "params": []
}
trace Information: 7 : {"id":2,"error":{"code":11050,"message":"bad_request"}}
trace Verbose: 7 : Received: {
  "jsonrpc": "2.0",
  "id": 2,
  "error": {
    "message": "bad_request",
    "code": 11050
  },
  "usIn": 1712668905757776,
  "usOut": 1712668905757838,
  "usDiff": 62,
  "testnet": false
}
trace Warning: 12 : Received error response for request 2: 11050 "bad_request"
```

![image](https://github.com/microsoft/vs-streamjsonrpc/assets/344148/32752101-b72a-4a90-8e1b-4a41af359fb4)

**Comment by AArnott** [comment]
What API are you using to invoke the RPC request? You can use the `JsonRpc.InvokeWithParameterObjectAsync` method to enforce an object be sent. If `null` as the parameter object still produces an array, that may be a bug. But it may be one you can workaround by passing in an object with no properties (or an empty dictionary IIRC).

**Reply by prodigy** [reply]
Thank you for your reply. `JsonRpc.InvokeWithParameterObjectAsync` sent an array as well just as you said. Possibly this is a bug.

I then tried using `.Attach<>()` with a custom server interface with the settings set to `ServerRequiresNamedArguments` to `true`.

This then didn't produce any arguments although they were passed as method arguments. It continued generating empty arrays, even with arguments present.

My current workaround actually is a bugfix I guess. After looking through the code of `StreamJsonRpc` and identifying the code that decides whether to send an object or an array, I came up with this class deriving from `JsonMessageFormatter`...

```csharp
public class ObjectOnlyJsonRpcFormatter : JsonMessageFormatter, IJsonRpcMessageFormatter
{
    void IJsonRpcMessageFormatter.Serialize(IBufferWriter<byte> bufferWriter, JsonRpcMessage message)
    {
        if (message is JsonRpcRequest { ArgumentsList.Count: 0 } request)
        {
            request.ArgumentsList = null;
            request.NamedArguments = new Dictionary<string, object?>();
            request.NamedArgumentDeclaredTypes = new Dictionary<string, Type>();
        }
        
        base.Serialize(bufferWriter, message);
    }
}
```
